Refactoring, simplifying payment processors

guzzilar opened the pull request:
> Note: This pull request requires PR #111 to be merged first.

### 1. Objective

Alipay, Credit Card, Internet Banking payments are currently behaving almost the same in many aspects (except for some part such as charge-creating and handling the charge response).

However, they are completely separated from each other while some part of code could have been shared among each payment method to reduce those redundant code and make the payment processor more easy to maintain.

### 2. Description of change

Previously, each of payment method's class will have its own `process_payment` which have the same structure that can be written as follows:
```php
public function process_payment( $order_id ) {
    // 1. Loading WooCommerce Order from the $order_id into $order

    try {
        // 2. Logging an initiating message to WooCommerce Order
        // 3. Gathering all data and create a new charge
        // 4. Logging that charge has been created, attaching OmiseCharge.id as WooCommerce Order's transaction id.
        // 5. Validating a charge result, result will be depending on its charge.status (successful, failed, pending).
    } catch (Exception $e) {
        // 6. Notify an error message to the checkout page (wc_add_notice())
        // 7. Then, logging an error message to WooCommerce Order.
    }
}
```

By the above structure, we can simply take those redundant code out and group it into one method.
```php
class Omise_Payment {
    public function process_payment( $order_id ) {
        // 1. Loading WooCommerce Order from the $order_id into $order

        try {
            // 2. Logging an initiating message to WooCommerce Order

            $charge = $this->charge( $order_id, $order );

            // 4. Logging that charge has been created, attaching OmiseCharge.id as WooCommerce Order's transaction id.

            return $this->result( $order_id, $order, $charge );
        } catch (Exception $e) {
            // 6. Notify an error message to the checkout page (wc_add_notice())
            // 7. Then, logging an error message to WooCommerce Order.
        }
    }
}
```

And let those payment processors concern only _how to create its own charge_ and _validating a charge result._

```php
class Omise_Payment_Alipay extends Omise_Payment {
    public function charge( $order_id, $order ) { }
    public function result( $order_id, $order, $charge ) { }
}

class Omise_Payment_Creditcard extends Omise_Payment {
    public function charge( $order_id, $order ) { }
    public function result( $order_id, $order, $charge ) { }
}

class Omise_Payment_Internetbanking extends Omise_Payment {
    public function charge( $order_id, $order ) { }
    public function result( $order_id, $order, $charge ) { }
}
```

By this approach, those payment processor classes would be much more smaller and have less responsibility, reducing unnecessary code & easy to maintain. Bonus point is also to decoupling the code, would make it more easy to test (either by manual test or unit test)

### 3. Quality assurance

**🔧 Environments:**

- **WordPress**: v5.0.3
- **WooCommerce**: v3.5.3
- **PHP version**: 5.6.30

**✏️ Details:**

**Because this pull request is refactoring on both 3 payment processor classes, it requires a serious test on both 3 payment methods' functionalities.**
Test details are as follows:

#### • Alipay Payment
There are 4 important cases that need to be tested.
**1. Making a new order with Alipay payment and mark as successful**
By placing an order as normal process, and choose **Alipay** as a payment method.
<img width="1552" alt="Screen Shot 2562-04-01 at 22 35 23" src="https://user-images.githubusercontent.com/2154669/55340134-450d3380-54ce-11e9-9a22-98297c2adbcf.png">

[Screenshot: Mark as success](https://user-images.githubusercontent.com/2154669/55340132-44749d00-54ce-11e9-8abb-c00f31b9e018.png), [Screenshot: order successfully](https://user-images.githubusercontent.com/2154669/55340133-44749d00-54ce-11e9-8e64-7f6db98a98a8.png)

**2. Making a new order with Alipay payment and mark as failed**
By placing an order as normal process, and choose **Alipay** as a payment method but mark the charge as failed.
<img width="1552" alt="Screen Shot 2562-04-01 at 22 35 00" src="https://user-images.githubusercontent.com/2154669/55340391-c664c600-54ce-11e9-94ee-d0f3e5c7f67b.png">

[Screenshot: Order failed](https://user-images.githubusercontent.com/2154669/55340365-bc42c780-54ce-11e9-9a32-89af9180d66e.png)

**3. Making a new order with Alipay payment and test in case of `charge.status` = `pending`.**
This test requires a tester to do hard-code a charge result to **"pending"** (as we don't have a proper way to mark a charge transaction as pending at the moment)

Add the following code to the line as in the below link:
```php
$charge['status']     = 'pending';
$charge['authorized'] = false;
$charge['paid']       = false;
```
https://github.com/omise/omise-woocommerce/blob/master/includes/gateway/class-omise-payment-alipay.php#L184

This to fake a charge result and force it into a race-condition where the result of a particular charge is not neither `successful` nor `failed`.

<img width="1552" alt="Screen Shot 2562-04-01 at 22 48 28" src="https://user-images.githubusercontent.com/2154669/55341712-49871b80-54d1-11e9-9366-24375028bf55.png">

**4. Make sure that Webhook feature still works properly.**

#### • Credit Card Payment
There are 7 important cases that need to be tested.
**1. Making a new order with `auth-only`.**
<img width="1552" alt="Screen Shot 2562-04-02 at 00 47 26" src="https://user-images.githubusercontent.com/2154669/55348230-f026e880-54e0-11e9-8702-4f8acc68362d.png">

[Screenshot: Order received](https://user-images.githubusercontent.com/2154669/55348252-fa48e700-54e0-11e9-9352-9ee6f285c1a4.png)

**2. Test capturing an authorized charge at WooCommerce Order Detail page.**
<img width="1552" alt="Screen Shot 2562-04-02 at 00 49 17" src="https://user-images.githubusercontent.com/2154669/55348360-3e3bec00-54e1-11e9-8dde-6402a1105666.png">

**3. Test refunding a captured charge at WooCommerce Order Detail page.**
<img width="1552" alt="Screen Shot 2562-04-02 at 00 51 01" src="https://user-images.githubusercontent.com/2154669/55348455-6e838a80-54e1-11e9-9948-49ec9d398df9.png">

**4. Making a new order with `auto-capture`.**
<img width="1552" alt="Screen Shot 2562-04-04 at 10 16 38" src="https://user-images.githubusercontent.com/2154669/55527333-c75e4900-56c2-11e9-9e01-f3642716c541.png">

**5. Making a new order and save a card**
<img width="1552" alt="Screen Shot 2562-04-04 at 10 23 28" src="https://user-images.githubusercontent.com/2154669/55527571-d5609980-56c3-11e9-868b-1dd566de4909.png">

<img width="1552" alt="Screen Shot 2562-04-04 at 10 24 12" src="https://user-images.githubusercontent.com/2154669/55527572-d5609980-56c3-11e9-996f-a6c04f5e995c.png">

**6. Making a new order and use a saved-card**

**7. Make sure that Webhook feature still works properly.**

#### • Internet Banking Payment
There are 4 important cases that need to be tested.
**1. Making a new order with Internet Banking payment and mark as successful**
By placing an order as normal process, and choose **Internet Banking** as a payment method.

<img width="1552" alt="Screen Shot 2562-04-01 at 23 49 00" src="https://user-images.githubusercontent.com/2154669/55344949-fa44e900-54d8-11e9-94f6-f8d273c4cbf0.png">

[Screenshot: Mark as success](https://user-images.githubusercontent.com/2154669/55344967-04ff7e00-54d9-11e9-8d95-f0fd0da28a2d.png), [Screenshot: Order detail page](https://user-images.githubusercontent.com/2154669/55344968-04ff7e00-54d9-11e9-84e4-746079a35b0d.png)

**2. Making a new order with Internet Banking payment and mark as failed**
By placing an order as normal process, and choose **Internet Banking** as a payment method but mark the charge as failed.
<img width="1552" alt="Screen Shot 2562-04-01 at 23 53 15" src="https://user-images.githubusercontent.com/2154669/55345169-73444080-54d9-11e9-9ba8-d8a0e8d57ede.png">

[Screenshot: Order detail page](https://user-images.githubusercontent.com/2154669/55345170-73dcd700-54d9-11e9-8cb3-1a516f0bf849.png)

**3. Making a new order with Internet Banking payment and test in case of `charge.status` = `pending`.**
This test requires a tester to do hard-code a charge result to **"pending"** (as we don't have a proper way to mark a charge transaction as pending at the moment)

Add the following code to the line as in the below link:
```php
$charge['status']     = 'pending';
$charge['authorized'] = false;
$charge['paid']       = false;
```
https://github.com/omise/omise-woocommerce/blob/master/includes/gateway/class-omise-payment-internetbanking.php#L199

This to fake a charge result and force it into a race-condition where the result of a particular charge is not neither `successful` nor `failed`.
<img width="1552" alt="Screen Shot 2562-04-02 at 00 41 55" src="https://user-images.githubusercontent.com/2154669/55347939-316ac880-54e0-11e9-8850-4e7aba22d51b.png">

**4. Make sure that Webhook feature still works properly.**

### 4. Impact of the change

- Some log messages are changed, the list of effected-messages are as follows:
[drafting]

### 5. Priority of change

Normal

### 6. Additional Notes

No

guzzilar commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,102 +61,60 @@ public function init_form_fields() {
+		 *
+		 * @return OmiseCharge|OmiseException
 		 */
-		public function process_payment( $order_id ) {
```
Note: This change might a bit hard to see because of GitHub trying to differ between the old code and a new one but basically, 
1. `process_payment` will be renaming to `charge`.

Then removing all `load order` and `add note to order` logic (moving it out to `Omise_Payment::process_payment`).

guzzilar commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,102 +61,60 @@ public function init_form_fields() {
 
-			try {
-				$metadata = apply_filters( 'omise_charge_params_metadata', array(), $order );
-				$charge = $this->sale( array(
```
Continue from https://github.com/omise/omise-woocommerce/pull/101/files#r259255182,
2. Moving this part up, instead of passing all parameters to `sale` method, this `Omise_Payment_Alipay::charge()` method will create a charge by itself, and return OmiseCharge object back to `Omise_Payment::process_payment`

guzzilar commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,102 +61,60 @@ public function init_form_fields() {
-
-				$order->add_order_note( sprintf( __( 'Omise: Charge (ID: %s) has been created', 'omise' ), $charge['id'] ) );
-
-				switch ( $charge['status'] ) {
```
Continue from https://github.com/omise/omise-woocommerce/pull/101/files#r259255705
3. This result-handler is moved to `Omise_Payment_Alipay::result()` method.

So, from the 3 parts of changes, we can
- Get rid of redundant code (that we repeated it in every payment processor class (Alipay, Credit Card, Internet Banking, and future payments, Bill Payment, Installment).

- Decoupling code, `charge creation` and `handle charge result` can be separated to its own method, easy to maintain and test.

guzzilar commented:
@jacstn @jonrandy I've finished this pull request. However, needs #111 to be merged first.

jonrandy commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,103 +61,47 @@ public function init_form_fields() {
+
+			return $this->payment_failed(
+				sprintf(
+					__( 'Please feel free to try submit your order again or contact our support team if you have any questions (Your temporary order id is \'%s\')', 'omise' ),
```
Please feel free to try submitting your order again, or contact our support team if you have any questions (Your temporary order id is \'%s\')

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+			$card_id = isset( $_POST['card_id'] ) ? wc_clean( $_POST['card_id'] ) : '';
+
+			if ( empty( $token ) && empty( $card_id ) ) {
+				throw new Exception( __( 'Please select an existing card or enter a new card information.', 'omise' ) );
```
Please select an existing card or enter new card information.

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+			// Saving card.
+			if ( isset( $_POST['omise_save_customer_card'] ) && empty( $card_id ) ) {
+				if ( empty( $token ) ) {
+					throw new Exception( __( 'Cannot process with a card that you\'re using. Please make sure that the card info is correct or contact our support team if you have any questions.', 'omise' ) );
```
Unable to process the card. Please make sure that the information is correct, or contact our support team if you have any questions.

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+						throw new Exception(
+							sprintf(
+								wp_kses(
+									__( 'Note that nothing wrong by you, this might be from our store issue.<br/><br/>Please feel free to try submit your order again or report our support team that you have found this problem (Your temporary order id is \'%s\')', 'omise' ),
```
Please note that you've done nothing wrong - this is likely an issue with our store.<br/><br/>Feel free to try submitting your order again, or report to our support team that you have found this problem (Your temporary order id is

jonrandy commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,103 +61,47 @@ public function init_form_fields() {
-					)
-				);
+			if ( 'pending' == $charge['status'] ) {
+				$order->add_order_note( sprintf( __( 'Omise: Redirecting buyer out to %s', 'omise' ), esc_url( $charge['authorize_uri'] ) ) );
```
Omise: Redirecting buyer to

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+			if ( $this->omise_3ds ) {
+				$order->add_order_note(
+					sprintf(
+						__( 'Omise: Processing with a 3-D Secure payment, redirecting buyer out to %s', 'omise' ),
```
Omise: Processing a 3-D Secure payment, redirecting buyer to %s

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+						$order->add_order_note(
+							sprintf(
+								wp_kses(
+									__( 'Omise: Payment processing.<br/>An amount %1$s %2$s has been authorized', 'omise' ),
```
Omise: Payment processing.<br/>An amount of %1$s %2$s has been authorized

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
 
-							break;
+			switch ( strtoupper( $this->payment_action ) ) {
+				case 'MANUAL_CAPTURE':
```
`'MANUAL_CAPTURE'` and `'AUTO_CAPTURE'` feel like they should be constants

jonrandy replied:
Other places too - same thing

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+						$order->add_order_note(
+							sprintf(
+								wp_kses(
+									__( 'Omise: Payment successful.<br/>An amount %1$s %2$s has been paid', 'omise' ),
```
Omise: Payment successful.<br/>An amount of %1$s %2$s has been paid

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+			}
+
+			if ( ! $success ) {
+				return $this->payment_failed( __( 'Note that your payment might already has been processed. Please contact our support team if you have any questions.', 'omise' ) );
```
Note that your payment may have already been processed. Please contact our support team if you have any questions.

jonrandy commented on includes/gateway/class-omise-payment-internetbanking.php:
```diff
@@ -77,102 +77,47 @@ public function payment_fields() {
-					'error'
-				);
+			if ( 'pending' == $charge['status'] ) {
+				$order->add_order_note( sprintf( __( 'Omise: Redirecting buyer out to %s', 'omise' ), esc_url( $charge['authorize_uri'] ) ) );
```
Omise: Redirecting buyer to %s

jonrandy commented on includes/gateway/class-omise-payment-internetbanking.php:
```diff
@@ -77,102 +77,47 @@ public function payment_fields() {
+
+			return $this->payment_failed(
+				sprintf(
+					__( 'Please feel free to try submit your order again or contact our support team if you have any questions (Your temporary order id is \'%s\')', 'omise' ),
```
Please feel free to try submitting your order again, or contact our support team if you have any questions (Your temporary order id is \'%s\')

jonrandy commented on includes/gateway/class-omise-payment.php:
```diff
@@ -267,6 +277,104 @@ public function sync_payment( $order ) {
+		wc_add_notice(
+			sprintf(
+				wp_kses(
+					__( 'We cannot process your payment.<br/>Note that nothing wrong by you, this might be from our store issue.<br/><br/>Please feel free to try submit your order again or report our support team that you have found this problem (Your temporary order id is \'%s\')', 'omise' ),
```
We have been unable to process your payment.<br/>Please note that you've done nothing wrong - this is likely an issue with our store.<br/><br/>Feel free to try submitting your order again, or report this problem to our support team (Your temporary order id is \'%s\')

jonrandy commented on includes/gateway/class-omise-payment.php:
```diff
@@ -267,6 +277,104 @@ public function sync_payment( $order ) {
+		wc_add_notice(
+			sprintf(
+				wp_kses(
+					__( 'Seems we cannot process your payment properly:<br/>%s', 'omise' ),
```
It seems we've been unable to process your payment properly

guzzilar commented:
@jonrandy sorry, seems I wiped all your comments out when I did force push.

_Note: All comments about response message are fixed at #112._

guzzilar commented:
@jonrandy I've addressed all places you commented. Can you help re-review again?

jonrandy commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,95 +61,47 @@ public function init_form_fields() {
+			);
 
-			$order->add_order_note( __( 'Omise: Processing a payment with Alipay solution..', 'omise' ) );
+			return OmiseCharge::create( array(
```
Any reason why using long array syntax? WPress guidelines say either is fine and suggest moving to short if no requirement for backwards compatibility

jonrandy replied:
https://make.wordpress.org/core/2019/03/26/coding-standards-updates-for-php-5-6/

guzzilar replied:
@jonrandy  To make its code style conform with the rest of our code base.
Omise-WooCommerce has been using long array syntax for entire project so I am avoiding making any confusion (we may replace all long array syntax to a short one later on if necessary)

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -555,8 +515,8 @@ public function callback() {
 
-				switch ( strtoupper( $this->payment_action ) ) {
-					case 'MANUAL_CAPTURE':
+				switch ( $this->payment_action ) {
```
`strtoupper` has been removed - is this because it wasn't actually needed? Or is this a potentially dangerous mistake?

guzzilar replied:
@jonrandy  It wasn't actually needed (I was afraid if it would raise any concerns so I put it there to make sure but in fact, this value has been set by its constant's value anyway, so it should always be a lower-case)

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,235 +169,192 @@ public function payment_fields() {
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( Omise_Charge::is_failed( $charge ) ) {
```
Sometimes `Omise_Charge::is_failed` is used, sometimes just a string check on charge status - should these all be standardised?

guzzilar replied:
👍 Agreed.
Actually, I have a plan to remove `Omise_Charge` and `includes/libraries/omise-plugin` later on (as it might not really related to this refactoring). Guess it will be more easy to review.

jonrandy commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,95 +61,47 @@ public function init_form_fields() {
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( 'failed' == $charge['status'] ) {
```
Hardcoded charge status smells bad (see also other comment on 'failure' checking)

guzzilar replied:
@jonrandy  Can change to a constant one at the class `Omise_Payment`.
Any idea on what name should it be?

jonrandy commented on includes/gateway/class-omise-payment-alipay.php:
```diff
@@ -61,95 +61,47 @@ public function init_form_fields() {
-						$e->getMessage()
-					)
-				);
+			if ( 'pending' == $charge['status'] ) {
```
Hardcoded charge status again

guzzilar replied:
@jonrandy Can change to a constant one at Omise_Payment class.
Any idea on what name should it be?

jonrandy replied:
Feels like it should be on `OmiseCharge`. Maybe `STATUS_FAILED` there?

jonrandy commented on includes/gateway/class-omise-payment-creditcard.php:
```diff
@@ -166,235 +169,192 @@ public function payment_fields() {
+			// Remove cart
+			WC()->cart->empty_cart();
+			return array (
+				'result'   => 'success',
```
Perhaps 'success' shouldn't be hardcoded?

guzzilar replied:
@jonrandy maybe, any idea on its constant name?

jonrandy replied:
Looks like something that is maybe already in WC? Is this return in a particular format that WC is expecting?

guzzilar replied:
@jonrandy yup, a payment provider who implemented this `process_payment` is expected to return `result` and `redirect` in an array format back.

Here is a usage-reference in WC class that executed `process_payment` method.
<img width="1055" alt="Screen Shot 2562-04-26 at 15 09 58" src="https://user-images.githubusercontent.com/2154669/56793638-a5299800-6836-11e9-8d54-5b0365f00f92.png">

jonrandy replied:
I guess if they don't have a constant for it, we can leave it

jonrandy commented on includes/gateway/class-omise-payment-internetbanking.php:
```diff
@@ -77,94 +77,47 @@ public function payment_fields() {
+				'amount'      => $this->format_amount_subunit( $order->get_total(), $order->get_order_currency() ),
+				'currency'    => $order->get_order_currency(),
+				'description' => apply_filters('omise_charge_params_description', 'WooCommerce Order id ' . $order_id, $order),
+				'source'      => array( 'type' => $_POST['omise-offsite'] ),
```
Potential danger with unsanitised `$_POST` value?

guzzilar replied:
👍 will sanitise it.

jonrandy replied:
ok

jonrandy commented on includes/gateway/class-omise-payment-internetbanking.php:
```diff
@@ -77,94 +77,47 @@ public function payment_fields() {
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( 'failed' == $charge['status'] ) {
```
Harcoded charge status

jonrandy replied:
Maybe should be a constant connect with Omise Charge somehow

jonrandy replied:
Since these charge statuses are 'part of' an OmiseCharge, it feels like the value for the constant should really come from there. Single source of truth and all that

guzzilar commented:
@jonrandy I've addressed the issues you've mentioned. Can you help review those again?
There are 2 related commits: 
- Replacing all hard-code payment status strings to constants. https://github.com/omise/omise-woocommerce/pull/101/commits/a71f1ec0483144b728b2f12cbb6e7149810396d5
- Internet Banking: sanitising POST parameter. https://github.com/omise/omise-woocommerce/pull/101/commits/8a671c0ffd95b7e9fc492b581a5b29561f0a40c7

In the meantime, I'll do retest those payment methods again according to the areas that I've made a change.

jonrandy commented:
I still have a bit of an issue with those constants

guzzilar commented:
@jonrandy regarding on your comment about constant should be coming from OmiseCharge.
It might be as you said, however, this topic could only be solved by implementing a constant or a method at the Omise-PHP library, which will take a long way to go and not quite really related to this pull request.

If it's not really a major thing, would it be possible to approve this one first and we can submit another pull request to enhance this plugin once those constants or methods are provided properly in Omise-PHP?
